qt/aqt/table.py

```diff
@@ -0,0 +1,1135 @@
+
+    def is_stale(self, threshold: float) -> bool:
+        return self.refreshed_at < threshold
+
+    @staticmethod
+    def generic(length: int, cell_text: str) -> CellRow:
+        return CellRow(
+            ((cell_text, False) for cell in range(length)),
+            BrowserRow.COLOR_DEFAULT,
+            "arial",
+            12,
+        )
+
+    @staticmethod
+    def placeholder(length: int) -> CellRow:
+        return CellRow.generic(length, "...")
+
+    @staticmethod
+    def deleted(length: int) -> CellRow:
+        return CellRow.generic(length, tr.browsing_row_deleted())
+
+
+def backend_color_to_aqt_color(color: BrowserRow.Color.V) -> Optional[Tuple[str, str]]:
+    if color == BrowserRow.COLOR_MARKED:
+        return colors.MARKED_BG
+    if color == BrowserRow.COLOR_SUSPENDED:
+        return colors.SUSPENDED_BG
+    if color == BrowserRow.COLOR_FLAG_RED:
+        return colors.FLAG1_BG
+    if color == BrowserRow.COLOR_FLAG_ORANGE:
+        return colors.FLAG2_BG
+    if color == BrowserRow.COLOR_FLAG_GREEN:
+        return colors.FLAG3_BG
+    if color == BrowserRow.COLOR_FLAG_BLUE:
+        return colors.FLAG4_BG
+    return None
+
+
+class DataModel(QAbstractTableModel):
+    def __init__(self, col: Collection, state: ItemState) -> None:
+        QAbstractTableModel.__init__(self)
+        self.col: Collection = col
+        self._state: ItemState = state
+        self._items: Sequence[Item] = []
+        self._rows: Dict[int, CellRow] = {}
+        self._last_refresh = 0.0
+        # serve stale content to avoid hitting the DB?
+        self._block_updates = False
+
+    # Row Object Interface
+    ######################################################################
+
+    # Get Rows
+
+    def get_cell(self, index: QModelIndex) -> Cell:
+        return self.get_row(index).cells[index.column()]
+
+    def get_row(self, index: QModelIndex) -> CellRow:
+        item = self.get_item(index)
+        if row := self._rows.get(item):
+            if not self._block_updates and row.is_stale(self._last_refresh):
+                # need to refresh
+                self._rows[item] = self._fetch_row_from_backend(item)
+                return self._rows[item]
+            # return row, even if it's stale
+            return row
+        if self._block_updates:
+            # blank row until we unblock
+            return CellRow.placeholder(self.len_columns())
+        # missing row, need to build
+        self._rows[item] = self._fetch_row_from_backend(item)
+        return self._rows[item]
+
+    def _fetch_row_from_backend(self, item: int) -> CellRow:
+        try:
+            row = CellRow(*self.col.browser_row_for_id(item))
+        except NotFoundError:
+            return CellRow.deleted(self.len_columns())
+        except Exception as e:
+            return CellRow.generic(self.len_columns(), str(e))
+
+        # fixme: hook needs state
+        gui_hooks.browser_did_fetch_row(item, row, self._state.active_columns)
```

Comment:
The hook should probably be passed the state here?

Reply:
Even just an is_note or similar boolean would be nice, and card_id will need renaming (card_or_note_id might be clearer than item_id?)